File: rugs/youtube.js
var youtube = new DuderRug("YouTube", "Do stuff with YouTube.");
youtube.storage = youtube.loadStorage();

youtube.getAPIKey = function() {
	if (this.storage.settings === undefined) {
		return false;
	} else if (this.storage.settings.api_key === undefined) {
		return false;
	}
	return this.storage.settings.api_key;
};

youtube.setAPIKey = function(key) {
	if (this.storage.settings === undefined) {
		this.storage.settings = {};
	}
	this.storage.settings.api_key = key;
	this.saveStorage(this.storage);
};

youtube.getPlaylistID = function(guildID) {
	if (this.storage[guildID] === undefined) {
		return false;
	} else if (this.storage[guildID].playlist === undefined) {
		return false;
	}
	return this.storage[guildID].playlist;
};

youtube.setPlaylistID = function(guildID, id) {
	if (this.storage[guildID] === undefined) {
		this.storage[guildID] = {};
	}
	this.storage[guildID].playlist = id;
	this.saveStorage(this.storage);
};

youtube.getVideosCache = function(guildID) {
	if (this.storage[guildID] === undefined) {
		return false;
	} else if (this.storage[guildID].videos === undefined) {
		return false;
	} else if (this.storage[guildID].videos.cache === undefined) {
		return false;
	}
	return this.storage[guildID].videos.cache;
};

youtube.setVideosCache = function(guildID, cache) {
	if (this.storage[guildID] === undefined) {
		this.storage[guildID] = {};
	}
	if (this.storage[guildID].videos === undefined) {
		this.storage[guildID].videos = {};
	}
	this.storage[guildID].videos.cache = cache;
	this.saveStorage(this.storage);
};

youtube.addCommand("yt", function(cmd) {
	var action = cmd.args.length > 1 ? cmd.args[1].toLowerCase() : "playlist";

	if (action === "setkey") {
		if (cmd.args.length === 3) {
			this.setAPIKey(cmd.args[2]);
			cmd.replyToAuthor("YouTube API key has been saved.");
			return;
		} else {
			cmd.replyToAuthor("usage: `setkey YOUR_API_KEY`");
			return;
		}
	}

	var api_key = this.getAPIKey();
	if (api_key === false) {
		cmd.replyToAuthor("no YouTube API key provided.");
		return;
	}

	if (action === "setplaylist") {
		if (cmd.args.length === 3) {
			this.setPlaylistID(cmd.guildID, cmd.args[2]);
			cmd.replyToAuthor("playlist `" + cmd.args[2] + "` has been saved.");
			return;
		} else {
			cmd.replyToAuthor("usage: `setplaylist PLAYLIST_ID`.");
			return;
		}
	} else if (action === "getplaylist") {
		var playlist = this.getPlaylistID(cmd.guildID);
		if (playlist === false) {
			cmd.replyToAuthor("the playlist hasn't been set.");
			return;
		} else {
			cmd.replyToAuthor("the current playlist is `" + playlist + "`.");
			return;
		}
	}

	var videos = this.getVideosCache(cmd.guildID);
	if (videos === false || action === "refreshplaylist") {
		this.dprint("Updating video cache");
		videos = [];

		var playlistID = this.getPlaylistID(cmd.guildID);
		if (playlistID === false) {
			cmd.replyToAuthor("the playlist hasn't been set.");
			return;
		}
		var baseurl = "https://www.googleapis.com/youtube/v3/playlistItems?playlistId={0}&maxResults=50&part=contentDetails&key={1}".format(
			playlistID,
			api_key
		);
		var url = baseurl;
		for (var i = 0; i < 10; i++) {
			var content = HTTP.get(4, url);
			var json = JSON.parse(content);
			for (var k in json.items) {
				var video = json.items[k];
				if (video.contentDetails !== undefined) {
					videos.push(video.contentDetails.videoId);
				}
			}

			if (json.nextPageToken !== undefined) {
				url = baseurl + "&pageToken=" + json.nextPageToken;
			} else {
				break;
			}
		}
		this.setVideosCache(cmd.guildID, videos);
		videos = this.getVideosCache(cmd.guildID);
		if (action === "refreshplaylist") {
			cmd.replyToAuthor("the playlist has been refreshed.");
		}
	} else {
		this.dprint("Using video cache");
	}

	var r = Math.getRandomInRange(0, videos.length);

	//this.print("amount of videos " + videos.length);
	cmd.replyToChannel("https://youtu.be/" + videos[r]);
});

File: rugs/youtube.go
package rugs

import (
	"encoding/json"
	"fmt"
	"os"

	"github.com/bwmarrin/discordgo"
)

func init() {
	Register(&YouTubeCommand{})
}

// YouTubeCommand implements the /youtube slash command
type YouTubeCommand struct {
	storage *Storage
}

func (c *YouTubeCommand) Name() string {
	return "youtube"
}

func (c *YouTubeCommand) Description() string {
	return "Get a random video from a YouTube playlist"
}

func (c *YouTubeCommand) Options() []*discordgo.ApplicationCommandOption {
	return []*discordgo.ApplicationCommandOption{
		{
			Type:        discordgo.ApplicationCommandOptionSubCommand,
			Name:        "random",
			Description: "Get a random video from the configured playlist",
		},
		{
			Type:        discordgo.ApplicationCommandOptionSubCommand,
			Name:        "setplaylist",
			Description: "Set the playlist ID",
			Options: []*discordgo.ApplicationCommandOption{
				{
					Type:        discordgo.ApplicationCommandOptionString,
					Name:        "playlist_id",
					Description: "YouTube playlist ID",
					Required:    true,
				},
			},
		},
		{
			Type:        discordgo.ApplicationCommandOptionSubCommand,
			Name:        "setkey",
			Description: "Set YouTube API key (owner only)",
			Options: []*discordgo.ApplicationCommandOption{
				{
					Type:        discordgo.ApplicationCommandOptionString,
					Name:        "key",
					Description: "YouTube Data API key",
					Required:    true,
				},
			},
		},
	}
}

func (c *YouTubeCommand) getStorage() *Storage {
	if c.storage == nil {
		c.storage = NewStorage("youtube")
		c.storage.Load()
	}
	return c.storage
}

func (c *YouTubeCommand) getAPIKey() string {
	storage := c.getStorage()
	if key, ok := storage.GetNested("settings", "api_key"); ok {
		if str, ok := key.(string); ok && str != "" {
			return str
		}
	}
	// Fallback to environment variable
	if key := os.Getenv("YOUTUBE_API_KEY"); key != "" {
		return key
	}
	return os.Getenv("GOOGLE_API_KEY")
}

func (c *YouTubeCommand) getPlaylistID() string {
	storage := c.getStorage()
	if id, ok := storage.GetNested("settings", "playlist_id"); ok {
		if str, ok := id.(string); ok {
			return str
		}
	}
	return ""
}

func (c *YouTubeCommand) Execute(ctx CommandContext) error {
	data := ctx.Interaction().ApplicationCommandData()
	if len(data.Options) == 0 {
		return ctx.ReplyEphemeral("Please specify a subcommand.")
	}

	subCmd := data.Options[0].Name

	switch subCmd {
	case "random":
		return c.handleRandom(ctx)
	case "setplaylist":
		return c.handleSetPlaylist(ctx)
	case "setkey":
		return c.handleSetKey(ctx)
	}

	return ctx.ReplyEphemeral("Unknown subcommand.")
}

func (c *YouTubeCommand) handleSetKey(ctx CommandContext) error {
	if !ctx.IsOwner() {
		return ctx.ReplyEphemeral("Only the bot owner can set the API key.")
	}

	key := ctx.GetString("key")
	storage := c.getStorage()
	storage.SetNested(key, "settings", "api_key")
	storage.Save()

	return ctx.ReplyEphemeral("YouTube API key set!")
}

func (c *YouTubeCommand) handleSetPlaylist(ctx CommandContext) error {
	playlistID := ctx.GetString("playlist_id")
	storage := c.getStorage()
	storage.SetNested(playlistID, "settings", "playlist_id")
	storage.Save()

	return ctx.Reply(fmt.Sprintf("Playlist set to: %s", playlistID))
}

func (c *YouTubeCommand) handleRandom(ctx CommandContext) error {
	apiKey := c.getAPIKey()
	if apiKey == "" {
		return ctx.ReplyEphemeral("YouTube API key not configured.")
	}

	playlistID := c.getPlaylistID()
	if playlistID == "" {
		return ctx.ReplyEphemeral("No playlist configured. Use `/youtube setplaylist` first.")
	}

	ctx.DeferReply()

	// Get playlist items
	url := fmt.Sprintf(
		"https://www.googleapis.com/youtube/v3/playlistItems?part=snippet&maxResults=50&playlistId=%s&key=%s",
		playlistID, apiKey)

	resp, err := ctx.HTTPGetString(10, url, nil)
	if err != nil {
		return ctx.FollowUp("Failed to fetch playlist.")
	}

	var result struct {
		Items []struct {
			Snippet struct {
				Title      string `json:"title"`
				ResourceID struct {
					VideoID string `json:"videoId"`
				} `json:"resourceId"`
				Thumbnails struct {
					Default struct {
						URL string `json:"url"`
					} `json:"default"`
				} `json:"thumbnails"`
			} `json:"snippet"`
		} `json:"items"`
	}

	if err := json.Unmarshal([]byte(resp), &result); err != nil || len(result.Items) == 0 {
		return ctx.FollowUp("No videos found in playlist.")
	}

	// Pick random video
	idx := RandomInRange(0, len(result.Items)-1)
	video := result.Items[idx]
	videoURL := fmt.Sprintf("https://www.youtube.com/watch?v=%s", video.Snippet.ResourceID.VideoID)

	embed := NewEmbed().
		SetTitle(fmt.Sprintf("🎬 %s", video.Snippet.Title)).
		SetURL(videoURL).
		SetThumbnail(video.Snippet.Thumbnails.Default.URL).
		SetColor(ColorRed).
		Build()

	return ctx.FollowUpEmbed(embed)
}
